Optionally use public images

smuda opened the pull request:
<!--
If this is a bug fix, make sure your description includes "Fixes: #xxxx", or
"Closes: #xxxx"

Please provide the following information:
-->
**- Description of the problem which is fixed/What is the use case**
Current build assumes the developer is a RH employee and uses images not publicly available.

Closes #268 

**- What I did**
- Updated Dockerfile to take images as arguments
- Updated current make target to include (the same) images as arguments when building docker images.
- Added documentation about building docker images from public images.

**- How to verify it**
Build with existing make targets:
```
make docker-build
make docker-buildx
```

Build with public images:
```
export BUILDER_IMAGE=registry.ci.openshift.org/openshift/release:golang-1.18
export TARGET_IMAGE=registry.ci.openshift.org/origin/4.10:base
make docker-build
make docker-buildx
```

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Added support for overriding images in Makefile.

bpradipt commented on Makefile:
```diff
@@ -60,6 +60,11 @@ else
+BUILDER_IMAGE_PUBLIC  = registry.ci.openshift.org/openshift/release:golang-1.18
+TARGET_IMAGE_PRIVATE  = registry.ci.openshift.org/ocp/4.11:base
+TARGET_IMAGE_PUBLIC   = registry.ci.openshift.org/origin/4.10:base
+
```
Can you please change the assignments to use `?=`. This will allows us to set these externally if required.
Also do you mind adding few lines in the doc to use the public targets .

smuda replied:
Sure!

smuda replied:
Fixed.

bpradipt commented:
@cpmeadors @jensfr ptal

bpradipt commented:
/ok-to-test

bpradipt reviewed (approved):
/lgtm
Thanks @smuda

cpmeadors commented:
I am concerned with these changes.  I understand the desire to have everything build-able for the community.  No one thinks we should have private RH images required.  The problem I see is that this is working around a problem that cause by internal RH build processes.  The proper solution is to use all publicly available resources here and fix downstream.

smuda commented:
I agree this is a workaround and it would be much better to have public images that is always used. Could you give me URL's for such images? I'd be happy to adjust the PR accordingly.

gkurz commented:
/hold

gkurz commented:
/hold

smuda commented:
Yes, I might have over-engineered a bit. 😀 I've simplified Makefile changes to simply allow overriding the images and added documentation to DEVELOPMENT.md with the appropriate public images.

The PR still does not address @cpmeadors [comment](https://github.com/openshift/sandboxed-containers-operator/pull/274#issuecomment-1470408450) about always having public images, but perhaps it's a step in the right direction?

bpradipt commented:
@gkurz ptal

gkurz commented:
> @gkurz ptal

Ack @bpradipt . Sorry @smuda  for the delay...

gkurz commented on docs/DEVELOPMENT.md:
```diff
@@ -22,6 +22,17 @@ In summary:
 
+### Using public images
+
+If you cannot login to registry.ci.openshift.org, a temporary solution is to use 
```
Nit: trailing space at end of line.

smuda replied:
I've removed trailing whitespace.

gkurz commented on Dockerfile:
```diff
@@ -1,5 +1,7 @@
 # Use OpenShift golang builder image
-FROM registry.ci.openshift.org/ocp/builder:rhel-8-golang-1.18-openshift-4.11 AS builder
+ARG BUILDER_IMAGE=${BUILDER_IMAGE:-registry.ci.openshift.org/ocp/builder:rhel-8-golang-1.18-openshift-4.11}
+ARG TARGET_IMAGE=${TARGET_IMAGE:-registry.ci.openshift.org/ocp/4.11:base}
```
Do we really need to duplicate the default values here ? AFAICT you'll get the default values from the Makefile.

@cpmeadors can you think of a case where this Dockerfile would be used standalone ?

smuda replied:
If my memory serves me right, it's required during automatic tests.

gkurz replied:
@smuda  That rings a bell indeed. Let's add a comment in both places stating that changes should be propagated to the other file as well.

smuda replied:
I've added a comment about keeping them in sync.

gkurz reviewed (approved):
Thanks @smuda !
